<a href="https://colab.research.google.com/github/rohitchoudhary55305-ctrl/Leve2Project2-Wine-Quality-Prediction-using-Machine-Learning/blob/main/WineQualityPredication.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [59]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
from matplotlib import style
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, classification_report, accuracy_score, ConfusionMatrixDisplay
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

In [60]:
import requests

# URL of the dataset
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/wine-quality/winequality-red.csv'

# Download the file
response = requests.get(url)
with open('winequality-red.csv', 'wb') as f:
    f.write(response.content)

print('winequality-red.csv downloaded successfully.')

winequality-red.csv downloaded successfully.


In [61]:
wine_df = pd.read_csv('winequality-red.csv', sep=';')
wine_df.head()

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality
0,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5
1,7.8,0.88,0.00,2.6,0.098,25.0,67.0,0.9968,3.20,0.68,9.8,5
2,7.8,0.76,0.04,2.3,0.092,15.0,54.0,0.9970,3.26,0.65,9.8,5
3,11.2,0.28,0.56,1.9,0.075,17.0,60.0,0.9980,3.16,0.58,9.8,6
4,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5


In [62]:
import pandas as pd

wine_df = pd.read_csv('winequality-red.csv', sep=';')
display(wine_df.head())

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality
0,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5
1,7.8,0.88,0.00,2.6,0.098,25.0,67.0,0.9968,3.20,0.68,9.8,5
2,7.8,0.76,0.04,2.3,0.092,15.0,54.0,0.9970,3.26,0.65,9.8,5
3,11.2,0.28,0.56,1.9,0.075,17.0,60.0,0.9980,3.16,0.58,9.8,6
4,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5


**EDA**

In [63]:
wine_df.shape

(1599, 12)

In [64]:
wine_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


In [65]:
wine_df.isnull().sum()

,0
fixed acidity,0
volatile acidity,0
citric acid,0
residual sugar,0
chlorides,0
free sulfur dioxide,0
total sulfur dioxide,0
density,0
pH,0
sulphates,0


In [66]:
wine_df.describe()

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality
count,1599.000000,1599.000000,1599.000000,1599.000000,1599.000000,1599.000000,1599.000000,1599.000000,1599.000000,1599.000000,1599.000000,1599.000000
mean,8.319637,0.527821,0.270976,2.538806,0.087467,15.874922,46.467792,0.996747,3.311113,0.658149,10.422983,5.636023
std,1.741096,0.179060,0.194801,1.409928,0.047065,10.460157,32.895324,0.001887,0.154386,0.169507,1.065668,0.807569
min,4.600000,0.120000,0.000000,0.900000,0.012000,1.000000,6.000000,0.990070,2.740000,0.330000,8.400000,3.000000
25%,7.100000,0.390000,0.090000,1.900000,0.070000,7.000000,22.000000,0.995600,3.210000,0.550000,9.500000,5.000000
50%,7.900000,0.520000,0.260000,2.200000,0.079000,14.000000,38.000000,0.996750,3.310000,0.620000,10.200000,6.000000
75%,9.200000,0.640000,0.420000,2.600000,0.090000,21.000000,62.000000,0.997835,3.400000,0.730000,11.100000,6.000000
max,15.900000,1.580000,1.000000,15.500000,0.611000,72.000000,289.000000,1.003690,4.010000,2.000000,14.900000,8.000000


In [67]:
wine_df['quality'].value_counts()

,count
quality,
5,681
6,638
7,199
4,53
8,18
3,10


In [ ]:
style.use('ggplot')
sns.countplot(wine_df['quality'])

<Axes: ylabel='count'>

In [ ]:
wine_df.hist(bins=100, figsize=(10,12))
plt.show()

In [ ]:
plt.figure(figsize=(10,7))
sns.heatmap(wine_df.corr(), annot=True)
plt.title('Correlation between the columns')
plt.show()

In [ ]:
wine_df.corr()['quality'].sort_values()

In [ ]:
sns.barplot(wine_df['quality'], wine_df['alcohol'])

Data **Processing**

In [ ]:
wine_df['quality'] = wine_df.quality.apply(lambda x:1 if x>=7 else 0)

In [ ]:
wine_df['quality'].value_counts()

In [ ]:
X = wine_df.drop('quality', axis=1)
y = wine_df['quality']

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.3, random_state=42)

In [ ]:
print("X_train ", X_train.shape)
print("y_train ", y_train.shape)
print("X_test ", X_test.shape)
print("y_test ", y_test.shape)

**Model Training**

logistic Regression model

In [ ]:
logreg = LogisticRegression()
logreg.fit(X_train, y_train)
logreg_pred = logreg.predict(X_test)
logreg_acc = accuracy_score(logreg_pred, y_test)
print("test accuracy is: {:.2f}%".format(logreg_acc*100))

In [ ]:
print(classification_report(y_test, logreg_pred))

In [ ]:
style.use('classic')
cm = confusion_matrix(y_test, logreg_pred, labels=logreg.classes_)
disp = ConfusionMatrixDisplay(confusion_matrix= cm, display_labels=logreg.classes_)
disp.plot()
print("TN: ", cm[0][0])
print("FN: ", cm[1][0])
print("TP: ", cm[1][1])
print("FP: ", cm[0][1])

**Decision Tree**

In [ ]:
dtree = DecisionTreeClassifier()
dtree.fit(X_train, y_train)
dtree_pred = dtree.predict(X_test)
dtree_acc = accuracy_score(dtree_pred, y_test)
print("Test accuracy: {:.2f}%".format(dtree_acc*100))

In [ ]:
print(classification_report(y_test, dtree_pred))

In [ ]:
style.use('classic')
cm = confusion_matrix(y_test, dtree_pred, labels=dtree.classes_)
disp = ConfusionMatrixDisplay(confusion_matrix= cm, display_labels=dtree.classes_)
disp.plot()
print("TN: ", cm[0][0])
print("FN: ", cm[1][0])
print("TP: ", cm[1][1])
print("FP: ", cm[0][1])


**Random Forest**

In [ ]:
rforest = RandomForestClassifier()
rforest.fit(X_train, y_train)
rforest_pred = rforest.predict(X_test)
rforest_acc = accuracy_score(rforest_pred, y_test)
print("Test accuracy: {:.2f}%".format(rforest_acc*100))

In [ ]:
print(classification_report(y_test, rforest_pred))

In [ ]:
style.use('classic')
cm = confusion_matrix(y_test, rforest_pred, labels=rforest.classes_)
disp = ConfusionMatrixDisplay(confusion_matrix= cm, display_labels=rforest.classes_)
disp.plot()
print("TN: ", cm[0][0])
print("FN: ", cm[1][0])
print("TP: ", cm[1][1])
print("FP: ", cm[0][1])

# Wine Quality Prediction Project: README

## Project Overview
This project aims to build a classification model to predict the quality of red wine based on its physicochemical properties. The goal is to classify wine into 'good quality' (quality >= 7) or 'bad quality' (quality < 7) using various machine learning algorithms.

## Data Source
The dataset used is the Wine Quality dataset (specifically, red wine) from the UCI Machine Learning Repository.

**Source URL:** `https://archive.ics.uci.edu/ml/machine-learning-databases/wine-quality/winequality-red.csv`

## Libraries Used
- `pandas` for data manipulation and analysis.
- `numpy` for numerical operations.
- `matplotlib.pyplot` for basic plotting.
- `seaborn` for advanced data visualization.
- `sklearn.model_selection` for splitting data into training and testing sets.
- `sklearn.metrics` for evaluating model performance (confusion matrix, classification report, accuracy score).
- `sklearn.preprocessing.StandardScaler` for feature scaling (though not used in the final models presented here, it was imported).
- `sklearn.linear_model.LogisticRegression` for Logistic Regression.
- `sklearn.tree.DecisionTreeClassifier` for Decision Tree.
- `sklearn.ensemble.RandomForestClassifier` for Random Forest.

## Project Steps and Analysis

### 1. Data Loading and Initial Inspection
- The `winequality-red.csv` file was downloaded from the specified URL using `requests`.
- The CSV file was loaded into a pandas DataFrame named `wine_df`.
- Initial data inspection was performed using:
    - `wine_df.head()`: Displayed the first few rows.
    - `wine_df.shape`: Showed the number of rows and columns (1599 rows, 12 columns).
    - `wine_df.info()`: Provided a summary of the DataFrame, including data types and non-null counts, confirming no missing values.
    - `wine_df.isnull().sum()`: Explicitly checked for null values, confirming no missing data.
    - `wine_df.describe()`: Generated descriptive statistics for numerical columns.

### 2. Exploratory Data Analysis (EDA)
- **Quality Distribution**: `wine_df['quality'].value_counts()` showed the distribution of wine quality ratings (ranging from 3 to 8), with most wines rated 5 or 6.
- **Visualizations**:
    - `sns.countplot(wine_df['quality'])`: Visualized the distribution of wine quality, reinforcing the imbalance towards quality levels 5 and 6.
    - `wine_df.hist(bins=100, figsize=(10,12))`: Generated histograms for all numerical features to understand their distributions.
    - `sns.heatmap(wine_df.corr(), annot=True)`: Displayed a correlation heatmap to identify relationships between features.
    - `wine_df.corr()['quality'].sort_values()`: Showed features sorted by their correlation with the `quality` target, highlighting `alcohol` and `sulphates` as positively correlated, and `volatile acidity`, `total sulfur dioxide`, and `density` as negatively correlated.
    - *Note on `sns.barplot(wine_df['quality'], wine_df['alcohol'])`*: An initial attempt to use `sns.barplot` directly with two positional arguments resulted in a `TypeError`. This could be corrected by specifying `x` and `y` arguments, e.g., `sns.barplot(x='quality', y='alcohol', data=wine_df)`.

### 3. Data Preprocessing
- **Target Binarization**: The `quality` column was transformed into a binary target variable:
    - Wines with `quality >= 7` were labeled as `1` (good quality).
    - Wines with `quality < 7` were labeled as `0` (bad quality).
    - `wine_df['quality'].value_counts()` after transformation showed 1382 wines labeled 0 and 217 labeled 1, indicating an imbalanced dataset.
- **Feature and Target Split**: The dataset was split into features `X` (all columns except `quality`) and target `y` (`quality` column).
- **Train-Test Split**: The data was further split into training (70%) and testing (30%) sets using `train_test_split` with `random_state=42` for reproducibility.
    - `X_train` shape: (1119, 11)
    - `y_train` shape: (1119,)
    - `X_test` shape: (480, 11)
    - `y_test` shape: (480,)

### 4. Model Training and Evaluation

#### 4.1. Logistic Regression
- A `LogisticRegression` model was trained on the `X_train` and `y_train` data.
- **Test Accuracy**: `86.67%`
- **Classification Report**:
```
              precision    recall  f1-score   support

           0       0.89      0.97      0.93       413
           1       0.56      0.22      0.32        67

    accuracy                           0.87       480
   macro avg       0.72      0.60      0.62       480
weighted avg       0.84      0.87      0.84       480
```
- **Confusion Matrix**:
    - TN: 401, FN: 52, TP: 15, FP: 12

#### 4.2. Decision Tree Classifier
- A `DecisionTreeClassifier` model was trained.
- **Test Accuracy**: `85.42%`
- **Classification Report**:
```
              precision    recall  f1-score   support

           0       0.93      0.89      0.91       413
           1       0.48      0.61      0.54        67

    accuracy                           0.85       480
   macro avg       0.71      0.75      0.73       480
weighted avg       0.87      0.85      0.86       480
```
- **Confusion Matrix**:
    - TN: 369, FN: 26, TP: 41, FP: 44

#### 4.3. Random Forest Classifier
- A `RandomForestClassifier` model was trained.
- **Test Accuracy**: `88.96%`
- **Classification Report**:
```
              precision    recall  f1-score   support

           0       0.92      0.95      0.94       413
           1       0.62      0.52      0.57        67

    accuracy                           0.89       480
   macro avg       0.77      0.74      0.75       480
weighted avg       0.88      0.89      0.89       480
```
- **Confusion Matrix**:
    - TN: 392, FN: 32, TP: 35, FP: 21

## Conclusion
Among the three models tested, the **Random Forest Classifier** achieved the highest test accuracy of `88.96%`. While all models performed reasonably well in classifying the majority class (quality 0), the recall for the minority class (quality 1) remains a challenge across all models due to the imbalanced nature of the dataset. Further steps could involve exploring techniques to handle imbalanced data (e.g., SMOTE, oversampling/undersampling) and hyperparameter tuning to improve the performance, especially for the good quality wines.